File: backend/rust/src/domain/value/transaction_type.rs
use core::fmt;
use core::str::FromStr;
use crate::domain::errors::DomainError;

#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
pub enum TransactionType {
    Deposit,
    Withdrawal,
    Send,
    Transfer,
    Payment
}

impl TransactionType {
    pub fn as_str(&self) -> &str {
        match self {
            TransactionType::Deposit    => "Deposit",
            TransactionType::Withdrawal => "Withdrawal",
            TransactionType::Send       => "Send",
            TransactionType::Transfer   => "Transfer",
            TransactionType::Payment    => "Payment",
        }
    }
}

impl fmt::Display for TransactionType {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(self.as_str())
    }
}

impl FromStr for TransactionType {
    type Err = DomainError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        // Case-insensitive parse normalized to the canonical Title-Case the DB
        // CHECK constraint requires (matches the C strcasecmp behaviour).
        match s.to_ascii_lowercase().as_str() {
            "deposit"    => Ok(TransactionType::Deposit),
            "withdrawal" => Ok(TransactionType::Withdrawal),
            "send"       => Ok(TransactionType::Send),
            "transfer"   => Ok(TransactionType::Transfer),
            "payment"    => Ok(TransactionType::Payment),
            _ => Err(DomainError::validation(format!("invalid TransactionType: {s}"))),
        }
    }
}

impl From<TransactionType> for String {
    fn from(value: TransactionType) -> Self {
        value.as_str().to_string()
    }
}

File: backend/c/domain/value/transaction_type.c
#include "../include/transaction_type.h"
#include <strings.h>
#include <string.h>

const char *transaction_type_as_str(TransactionType t) {
    switch (t) {
        case TRANSACTION_TYPE_DEPOSIT:    return "Deposit";
        case TRANSACTION_TYPE_WITHDRAWAL: return "Withdrawal";
        case TRANSACTION_TYPE_SEND:       return "Send";
        case TRANSACTION_TYPE_TRANSFER:   return "Transfer";
        case TRANSACTION_TYPE_PAYMENT:    return "Payment";
        default:                          return "Unknown";
    }
}

bool transaction_type_from_str(const char *s, TransactionType *out, DomainError *err) {
    if (!s || !out) {
        if (err) *err = domain_error_validation("transaction type string is null");
        return false;
    }

    if (!strcasecmp(s, "Deposit")) {
        *out = TRANSACTION_TYPE_DEPOSIT;
    } else if (!strcasecmp(s, "Withdrawal")) {
        *out = TRANSACTION_TYPE_WITHDRAWAL;
    } else if (!strcasecmp(s, "Send")) {
        *out = TRANSACTION_TYPE_SEND;
    } else if (!strcasecmp(s, "Transfer")) {
        *out = TRANSACTION_TYPE_TRANSFER;
    } else if (!strcasecmp(s, "Payment")) {
        *out = TRANSACTION_TYPE_PAYMENT;
    } else {
        if (err) *err = domain_error_validation("Invalid transaction type");
        return false;
    }

    if (err) *err = domain_error_ok();
    return true;
}
